### ISSUE TITLE
Registering a SailConnectionListener on a NotifyingSailConnection breaks transaction semantics, resulting in data loss

### ISSUE THREAD

**Issue by manuelfiorelli** [eclipse-rdf4j/rdf4j#5221]
### Current Behavior

I have a `MemoryStore` that contains only one triple _t_.
Then I run a _SPARQL Update_ on this repository, which deletes the triple _t_ and inserts the same triple.
After the update, the triple store still contains only the triple _t_, as expected.

When I register a `SailConnectionListener` on the sail connection I use to execute the update, I get a different result: the triple _t_ is deleted (in fact, I see a notification for the removal of the triple and not for its addition).

### Expected Behavior

 I expected that the triple _t_ would not be deleted regardless of the registration of a sail connection listener.

### Steps To Reproduce

To reproduce the issue, run the following program by toggling the value of the `useListener` variable between `false` and  `true`.
This is the version of Java in use:
```
java --version
openjdk 21.0.4 2024-07-16 LTS
OpenJDK Runtime Environment Temurin-21.0.4+7 (build 21.0.4+7-LTS)
OpenJDK 64-Bit Server VM Temurin-21.0.4+7 (build 21.0.4+7-LTS, mixed mode, sharing)
```

The output should look like this (the part between the two sets of dashes is a dump of the repository):

* `useListener == false`
```
----

<http://example.org/batman> a <http://example.org/Character> .
----
```
* `useListener == true`
```
Removed: (http://example.org/batman, http://www.w3.org/1999/02/22-rdf-syntax-ns#type, http://example.org/Character) [null]
----
----
```
```java
import org.eclipse.rdf4j.model.Statement;
import org.eclipse.rdf4j.model.util.Values;
import org.eclipse.rdf4j.model.vocabulary.RDF;
import org.eclipse.rdf4j.repository.sail.SailRepository;
import org.eclipse.rdf4j.rio.RDFFormat;
import org.eclipse.rdf4j.rio.Rio;
import org.eclipse.rdf4j.sail.NotifyingSailConnection;
import org.eclipse.rdf4j.sail.SailConnectionListener;
import org.eclipse.rdf4j.sail.memory.MemoryStore;

public class Scratch {
    public static void main(String[] args) {
        var useListener = true;

        var repo = new SailRepository(new MemoryStore());
        try {
            repo.init();

            var subject = Values.iri("http://example.org/batman");
            var clazz = Values.iri("http://example.org/Character");

            try (var con = repo.getConnection()) {
                con.add(subject, RDF.TYPE, clazz);
            }

            try (var con = repo.getConnection()) {
                con.begin();

                if (useListener) {
                    ((NotifyingSailConnection) con.getSailConnection()).addConnectionListener(new SailConnectionListener() {
                        @Override
                        public void statementAdded(Statement st) {
                            System.out.println("Added: " + st);
                        }

                        @Override
                        public void statementRemoved(Statement st) {
                            System.out.println("Removed: " + st);
                        }
                    });
                }

                var op = con.prepareUpdate("""
                        PREFIX ex: <http://example.org/>
                        
                        DELETE {
                            ex:batman a ex:Character .
                        }
                        INSERT {
                            ex:batman a ex:Character .
                        }
                        WHERE {
                        }
                        """);
                op.execute();
                con.commit();
            }

            try (var con = repo.getConnection()) {
                System.out.println("----");
                con.export(Rio.createWriter(RDFFormat.TURTLE, System.out));
                System.out.println("----");
            }

        } finally {
            repo.shutDown();
        }
    }
}
```

### Version

4.3.15 (also in 5.1.0)

### Are you interested in contributing a solution yourself?

No

**Comment by hmottestad** [comment]
Thanks for reporting this @manuelfiorelli !

The code was very helpful, and I've figured out what's going on. Both the DELETE and INSERT operations run on a dataset that doesn't reflect any changes until both operations are complete. This follows the semantics in the SPARQL standard. 

There are two issues. The first is that the code used to check if a statement already exists before notifying the addition/removal doesn't see the changes from the DELETE operation. The second issue is that the code that actually adds the statement is "optimised" to only add the statement if it doesn't already exist.

**Comment by manuelfiorelli** [comment]
Thank you @hmottestad for taking care of the issue. I also appreciated the detailed explanation of the defects that caused the reported problem.

**Comment by hmottestad** [comment]
A fix for this will be available in RDF4J 5.1.1.
